File: apps/api/plane/db/models/workflow_transition.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""
Governed multi-state workflows with approvals - see
docs/feature-specs/06-automation-workflow-sla.md ("Workflows gouvernés
multi-états avec approbations", section 4) in plane-selfhost.

PHASE 1 OF 2 (backend-only, per this session's build convention - see the
other category 6 features for the same split): this module defines the
complete data model and, in `plane/utils/workflow_transition_engine.py`,
the complete validation/execution engine plus every endpoint that manages
the engine's own configuration and approval flow. It deliberately does
NOT wire enforcement into the real `Issue.state` mutation paths (unitary
update, bulk update, public/service-token API) - a phase 2 change inserts
the `evaluate_transition`/`execute_allowed_transition` calls at those
existing call sites. See this feature's delivery report for the exact
list of files/call sites phase 2 needs to touch.

Six new tables, exactly as specified:
- `WorkflowTransition` - the (project, issue_type, from_state, to_state)
  graph edge itself.
- `WorkflowTransitionApprover` - per-transition role/member restriction
  and/or approval requirement (see `WorkflowTransitionApprover`'s own
  docstring for the resolved "approval_required" semantics - the spec's
  own text is genuinely underspecified here, see
  plane/utils/workflow_transition_engine.py's module docstring for the
  full reasoning, mirroring the rigor of plane/utils/sla_engine.py's own
  resolved-ambiguity documentation).
- `WorkflowTransitionCondition` - AND-combined preconditions.
- `WorkflowTransitionAction` - ordered post-transition actions.
- `IssueTransitionApprovalRequest` / `IssueTransitionApproval` - the
  approval flow itself.
- `WorkflowTransitionAuditLog` - every evaluated attempt (allowed, denied,
  pending_approval), kept separate from `IssueActivity` per the spec's own
  "pour ne pas polluer IssueActivity avec des événements refusés" (this is
  the spec's own reasoning, restated directly from the "Implications sur
  le modèle de données" section, not an invented justification).

SCOPE NARROWING - `WorkflowTransitionCondition.REQUIRED_FIELDS_FILLED`:
the spec's exigence 8 says "tous les champs obligatoires configurés sont
renseignés", which in Linear's own product refers to a configurable
custom-field/issue-property system. No such system exists anywhere in
this codebase (confirmed: no `IssueProperty`/`IssuePropertyValue` model,
nothing resembling one) - seen already this session for Category 5/6's
other features. This condition type is scoped down to a configurable list
of STANDARD `Issue` field names instead (`config: {"field_names": [...]}`,
e.g. `["description_html", "estimate_point", "target_date",
"assignee_ids", "label_ids"]`), checked for non-null/non-empty - see
`plane/utils/workflow_transition_engine.py::_check_required_fields_filled`.
This is a deliberate, honest adaptation to what actually exists, not a
silent guess.
"""

# Django imports
from django.db import models

# Module imports
from plane.db.models.base import BaseModel
from plane.db.models.project import ProjectBaseModel


class WorkflowTransition(ProjectBaseModel):
    """
    One allowed (or, once created for a (project, issue_type) pair,
    explicitly enumerated) `from_state -> to_state` edge - exigences 1-4.

    `issue_type=None` means "applies to all issue types in this project
    that don't have a more specific rule" (exigence 1's "ou pour 'tous les
    types' si aucune règle spécifique n'existe") - see
    plane/utils/workflow_transition_engine.py::evaluate_transition for the
    exact fallback/matching order.

    `from_state=None` means "transition from issue creation" (exigence 4 -
    the initial-state transition can be restricted exactly like any other).

    BACKWARD COMPATIBILITY (exigence 2 - the single most important
    correctness property of this whole feature): a (project, issue_type)
    pair with ZERO `WorkflowTransition` rows at all is an OPEN graph -
    every transition remains allowed, exactly as before this feature
    existed. The graph only becomes CLOSED for a pair once at least one
    row exists for it (exigence 3) - see the engine module for how "at
    least one row exists" is checked (both the type-specific and
    type-null rows count towards this).
    """

    issue_type = models.ForeignKey(
        "db.IssueType",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="workflow_transitions",
    )
    from_state = models.ForeignKey(
        "db.State",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="workflow_transitions_from",
    )
    to_state = models.ForeignKey(
        "db.State",
        on_delete=models.CASCADE,
        related_name="workflow_transitions_to",
    )
    is_active = models.BooleanField(default=True)

    class Meta:
        verbose_name = "Workflow Transition"
        verbose_name_plural = "Workflow Transitions"
        db_table = "workflow_transitions"
        ordering = ("-created_at",)
        # `deleted_at`-scoped, matching every other soft-deletable model in
        # this codebase with a natural-key uniqueness requirement (State,
        # ProjectIssueType, IssueSLA, IssueRelation) - `.delete()` on any
        # `AuditModel` subclass defaults to a SOFT delete (see
        # SoftDeleteModel.delete() in plane/db/mixins.py), so a plain
        # (without deleted_at) constraint would permanently block
        # recreating a transition for the same (issue_type, from_state,
        # to_state) tuple after deleting one - unacceptable for a
        # config screen where delete-then-recreate is a normal workflow.
        # NOTE: like `IssueSLA.sla_policy` (also nullable, also part of a
        # unique_together), Postgres treats NULL as distinct-from-itself in
        # a unique index, so two rows that are both e.g. `issue_type=None,
        # from_state=None, to_state=X` are not actually deduplicated by
        # this constraint - an accepted, pre-existing limitation in this
        # codebase's convention, not one newly introduced here.
        unique_together = ["project", "issue_type", "from_state", "to_state", "deleted_at"]
        constraints = [
            models.UniqueConstraint(
                fields=["project", "issue_type", "from_state", "to_state"],
                condition=models.Q(deleted_at__isnull=True),
                name="workflow_transition_unique_project_type_from_to_when_deleted_at_null",
            )
        ]

    def __str__(self):
        return f"{self.from_state_id} -> {self.to_state_id} <{self.project_id}>"


class WorkflowTransitionApprover(BaseModel):
    """
    A per-transition role or named-member entry - exigence 5/6.

    RESOLVED AMBIGUITY (see plane/utils/workflow_transition_engine.py's
    module docstring for the full reasoning, restated briefly here):
    `approval_required=False` rows are "these roles/members can execute
    this transition immediately, no approval needed for them specifically".
    `approval_required=True` rows are "these roles/members are valid
    approvers when a request DOES need approval". A transition with
    approver rows at all is role/member-restricted (exigence 5); whether a
    given actor within that restriction needs to go through the approval
    flow depends on which *kind* of row(s) they match, evaluated per-actor
    at evaluation time - see `evaluate_transition`.

    Constraint (app-level only, not a DB CHECK - same convention as
    `WorkflowTransition`'s own DB-level unique constraint being the only
    hard constraint, and e.g. `SLAPolicy`'s "at least one of
    response_time_minutes/resolution_time_minutes" being serializer-level
    only): exactly one of `member`/`role` must be set. Enforced in
    `app/views/workflow_transition/base.py`.
    """

    transition = models.ForeignKey(WorkflowTransition, on_delete=models.CASCADE, related_name="approvers")
    member = models.ForeignKey(
        "db.ProjectMember",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="workflow_transition_approvals",
    )
    # Raw `plane.app.permissions.ROLE` enum value (e.g. 20 for Admin, 15 for
    # Member) - Guest (5) is intentionally still a valid stored value here
    # (the DB doesn't enforce which roles are meaningful) but exigence 5's
    # "Guest ne peut jamais exécuter une transition restreinte" is enforced
    # at evaluation time in the engine, not by restricting what can be
    # configured.
    role = models.IntegerField(null=True, blank=True)
    approval_required = models.BooleanField(default=False)

    class Meta:
        verbose_name = "Workflow Transition Approver"
        verbose_name_plural = "Workflow Transition Approvers"
        db_table = "workflow_transition_approvers"
        ordering = ("-created_at",)

    def __str__(self):
        return f"member={self.member_id} role={self.role} required={self.approval_required} <{self.transition_id}>"


class WorkflowTransitionCondition(BaseModel):
    """
    An AND-combined precondition on a transition - exigence 3/8. See this
    module's docstring for the `REQUIRED_FIELDS_FILLED` scope narrowing to
    standard `Issue` fields (no custom-field system exists in this
    codebase).
    """

    CONDITION_TYPE_CHOICES = (
        ("SUB_ISSUES_CLOSED", "Sub-issues closed"),
        ("REQUIRED_FIELDS_FILLED", "Required fields filled"),
        ("NO_UNRESOLVED_BLOCKERS", "No unresolved blockers"),
        ("LABEL_PRESENT", "Label present"),
        ("LABEL_ABSENT", "Label absent"),
    )

    transition = models.ForeignKey(WorkflowTransition, on_delete=models.CASCADE, related_name="conditions")
    condition_type = models.CharField(max_length=30, choices=CONDITION_TYPE_CHOICES)
    # Shapes: REQUIRED_FIELDS_FILLED -> {"field_names": [...]} (see module
    # docstring); LABEL_PRESENT/LABEL_ABSENT -> {"label_id": "..."};
    # SUB_ISSUES_CLOSED/NO_UNRESOLVED_BLOCKERS need no config.
    config = models.JSONField(default=dict, blank=True)

    class Meta:
        verbose_name = "Workflow Transition Condition"
        verbose_name_plural = "Workflow Transition Conditions"
        db_table = "workflow_transition_conditions"
        ordering = ("-created_at",)

    def __str__(self):
        return f"{self.condition_type} <{self.transition_id}>"


class WorkflowTransitionAction(BaseModel):
    """
    An ordered post-transition action - exigence 9. Best-effort sequential
    application, mirroring `plane/utils/workflow_rule_engine.py`'s
    `_execute_rule_actions` exactly (see
    `plane/utils/workflow_transition_engine.py::_execute_transition_actions`).
    """

    ACTION_TYPE_CHOICES = (
        ("WEBHOOK", "Webhook"),
        ("SYSTEM_COMMENT", "System comment"),
        ("NOTIFY_ASSIGNEE", "Notify assignee"),
        ("NOTIFY_WATCHERS", "Notify watchers"),
        ("ADD_LABEL", "Add label"),
        ("REMOVE_LABEL", "Remove label"),
        ("ASSIGN_MEMBER", "Assign member"),
    )

    transition = models.ForeignKey(WorkflowTransition, on_delete=models.CASCADE, related_name="actions")
    action_type = models.CharField(max_length=30, choices=ACTION_TYPE_CHOICES)
    # Shapes: SYSTEM_COMMENT -> {"comment_template": "..."} (same
    # {{issue.identifier}}/{{issue.title}}/{{actor.display_name}} token
    # style as WorkflowAction.action_config's POST_COMMENT, see
    # plane/utils/workflow_rule_engine.py::_resolve_template);
    # ADD_LABEL/REMOVE_LABEL -> {"label_id": "..."}; ASSIGN_MEMBER ->
    # {"member_id": "..."}; WEBHOOK/NOTIFY_ASSIGNEE/NOTIFY_WATCHERS need no
    # config.
    config = models.JSONField(default=dict, blank=True)
    sort_order = models.PositiveIntegerField(default=0)

    class Meta:
        verbose_name = "Workflow Transition Action"
        verbose_name_plural = "Workflow Transition Actions"
        db_table = "workflow_transition_actions"
        ordering = ("sort_order",)

    def __str__(self):
        return f"{self.action_type} <{self.transition_id}>"


class IssueTransitionApprovalRequest(ProjectBaseModel):
    """
    A pending (or resolved) request to execute a transition that requires
    approval - exigence 6/7. `status` starts `PENDING`; a rejection sets
    `REJECTED` and, per exigence 7, must be re-submitted (a fresh request
    created) rather than re-evaluated in place.
    """

    STATUS_CHOICES = (
        ("PENDING", "Pending"),
        ("APPROVED", "Approved"),
        ("REJECTED", "Rejected"),
        ("CANCELLED", "Cancelled"),
    )

    issue = models.ForeignKey("db.Issue", on_delete=models.CASCADE, related_name="transition_approval_requests")
    transition = models.ForeignKey(WorkflowTransition, on_delete=models.CASCADE, related_name="approval_requests")
    requested_by = models.ForeignKey(
        "db.User", on_delete=models.SET_NULL, null=True, related_name="requested_transition_approvals"
    )
    status = models.CharField(max_length=20, choices=STATUS_CHOICES, default="PENDING", db_index=True)

    class Meta:
        verbose_name = "Issue Transition Approval Request"
        verbose_name_plural = "Issue Transition Approval Requests"
        db_table = "issue_transition_approval_requests"
        ordering = ("-created_at",)

    def __str__(self):
        return f"{self.issue_id} {self.transition_id} {self.status}"


class IssueTransitionApproval(BaseModel):
    """One approver's decision on an `IssueTransitionApprovalRequest`."""

    DECISION_CHOICES = (("APPROVED", "Approved"), ("REJECTED", "Rejected"))

    approval_request = models.ForeignKey(
        IssueTransitionApprovalRequest, on_delete=models.CASCADE, related_name="approvals"
    )
    approver = models.ForeignKey("db.User", on_delete=models.SET_NULL, null=True, related_name="transition_approvals")
    decision = models.CharField(max_length=20, choices=DECISION_CHOICES)
    comment = models.TextField(blank=True, default="")
    responded_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        verbose_name = "Issue Transition Approval"
        verbose_name_plural = "Issue Transition Approvals"
        db_table = "issue_transition_approvals"
        ordering = ("-responded_at",)

    def __str__(self):
        return f"{self.approval_request_id} {self.decision} <{self.approver_id}>"


class WorkflowTransitionAuditLog(BaseModel):
    """
    Every evaluated transition attempt - allowed, denied, or sent to
    approval - exigence 12/13. Kept separate from `IssueActivity` (the
    spec's own "Implications sur le modèle de données" reasoning: "pour ne
    pas polluer IssueActivity avec des événements refusés" - a denied
    attempt never touched the issue at all, so it has no place in a log of
    actual changes).

    `transition` is nullable for the OPEN-graph case (exigence 2) - zero
    rules matched at all because the (project, issue_type) pair has no
    configuration, so there is no specific `WorkflowTransition` to
    reference even though the outcome is still logged as `ALLOWED`.
    """

    OUTCOME_CHOICES = (("ALLOWED", "Allowed"), ("DENIED", "Denied"), ("PENDING_APPROVAL", "Pending approval"))

    issue = models.ForeignKey("db.Issue", on_delete=models.CASCADE, related_name="workflow_transition_audit_logs")
    transition = models.ForeignKey(
        WorkflowTransition, on_delete=models.SET_NULL, null=True, blank=True, related_name="audit_logs"
    )
    actor = models.ForeignKey(
        "db.User", on_delete=models.SET_NULL, null=True, related_name="workflow_transition_audit_logs"
    )
    from_state = models.ForeignKey(
        "db.State", on_delete=models.SET_NULL, null=True, blank=True, related_name="workflow_audit_logs_from"
    )
    to_state = models.ForeignKey(
        "db.State", on_delete=models.SET_NULL, null=True, blank=True, related_name="workflow_audit_logs_to"
    )
    outcome = models.CharField(max_length=20, choices=OUTCOME_CHOICES, db_index=True)
    denial_reason = models.TextField(null=True, blank=True)
    # `created_at` (auto_now_add=True) already comes from BaseModel's
    # TimeAuditModel mixin - not redeclared here.

    class Meta:
        verbose_name = "Workflow Transition Audit Log"
        verbose_name_plural = "Workflow Transition Audit Logs"
        db_table = "workflow_transition_audit_logs"
        ordering = ("-created_at",)
        indexes = [models.Index(fields=["issue", "created_at"], name="wf_transition_audit_issue_idx")]

    def __str__(self):
        return f"{self.issue_id} {self.outcome} <{self.transition_id}>"

File: packages/types/src/workflow-transition.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

// Types for governed multi-state workflows with approvals - see
// docs/feature-specs/06-automation-workflow-sla.md ("Workflows gouvernes
// multi-etats avec approbations", section 4) in plane-selfhost, and the
// backend half of this feature: apps/api/plane/db/models/workflow_transition.py,
// apps/api/plane/app/serializers/workflow_transition.py,
// apps/api/plane/app/views/workflow_transition/base.py,
// apps/api/plane/utils/workflow_transition_engine.py. Deliberately named
// `TWorkflowTransition*` (not `TWorkflowRule*`, which already exists for the
// sibling, unrelated trigger/condition/action automation engine built for
// category 6 feature 1 - a `WorkflowTransition` is a from-state -> to-state
// graph edge with approvers, not a trigger-fired rule).

/** `WorkflowTransition.approvers` row - exactly one of `member`/`role` is
 * set (enforced server-side, not by this type). See the model's own
 * docstring for the resolved "approval_required" semantics: a `false` row
 * grants direct-execution rights to that member/role even if another row on
 * the same transition has `approval_required=true`; a `true` row only makes
 * that member/role a valid APPROVER for someone else's pending request. */
export type TWorkflowTransitionApprover = {
  id?: string;
  member: string | null;
  /** Raw `plane.app.permissions.ROLE` value - `EUserProjectRoles.ADMIN`
   * (20) / `MEMBER` (15) / `GUEST` (5). */
  role: number | null;
  approval_required: boolean;
};

export type TWorkflowTransitionConditionType =
  | "SUB_ISSUES_CLOSED"
  | "REQUIRED_FIELDS_FILLED"
  | "NO_UNRESOLVED_BLOCKERS"
  | "LABEL_PRESENT"
  | "LABEL_ABSENT";

/** Only `REQUIRED_FIELDS_FILLED`/`LABEL_PRESENT`/`LABEL_ABSENT` carry any
 * config - `SUB_ISSUES_CLOSED`/`NO_UNRESOLVED_BLOCKERS` need none. Scoped to
 * standard `Issue` fields (no custom-field/issue-property system exists in
 * this codebase - see the backend model's own module docstring). */
export type TWorkflowTransitionConditionConfigRequiredFields = { field_names: string[] };
export type TWorkflowTransitionConditionConfigLabel = { label_id: string };
export type TWorkflowTransitionConditionConfig =
  | TWorkflowTransitionConditionConfigRequiredFields
  | TWorkflowTransitionConditionConfigLabel
  | Record<string, never>;

export type TWorkflowTransitionCondition = {
  id?: string;
  condition_type: TWorkflowTransitionConditionType;
  config: TWorkflowTransitionConditionConfig;
};

/** Standard `Issue` fields `REQUIRED_FIELDS_FILLED` may reference - mirrors
 * `_M2M_FIELD_ACCESSORS` plus the plain-field branch in
 * apps/api/plane/utils/workflow_transition_engine.py exactly. */
export const WORKFLOW_TRANSITION_REQUIRED_FIELD_OPTIONS = [
  { value: "description_html", label: "Description" },
  { value: "estimate_point", label: "Estimate" },
  { value: "target_date", label: "Due date" },
  { value: "start_date", label: "Start date" },
  { value: "assignee_ids", label: "Assignees" },
  { value: "label_ids", label: "Labels" },
] as const;

export type TWorkflowTransitionActionType =
  | "WEBHOOK"
  | "SYSTEM_COMMENT"
  | "NOTIFY_ASSIGNEE"
  | "NOTIFY_WATCHERS"
  | "ADD_LABEL"
  | "REMOVE_LABEL"
  | "ASSIGN_MEMBER";

/** Only `SYSTEM_COMMENT`/`ADD_LABEL`/`REMOVE_LABEL`/`ASSIGN_MEMBER` carry any
 * config - `WEBHOOK`/`NOTIFY_ASSIGNEE`/`NOTIFY_WATCHERS` need none. */
export type TWorkflowTransitionActionConfigSystemComment = { comment_template: string };
export type TWorkflowTransitionActionConfigLabel = { label_id: string };
export type TWorkflowTransitionActionConfigAssignMember = { member_id: string };
export type TWorkflowTransitionActionConfig =
  | TWorkflowTransitionActionConfigSystemComment
  | TWorkflowTransitionActionConfigLabel
  | TWorkflowTransitionActionConfigAssignMember
  | Record<string, never>;

export type TWorkflowTransitionAction = {
  id?: string;
  action_type: TWorkflowTransitionActionType;
  config: TWorkflowTransitionActionConfig;
  sort_order: number;
};

export type TWorkflowTransition = {
  id: string;
  workspace_id: string;
  project_id: string;
  /** `null` = "applies to all issue types in this project that don't have a
   * more specific rule" - see the backend model's docstring. */
  issue_type: string | null;
  /** `null` = "transition from issue creation" - not enforced anywhere yet
   * (phase 2 deliberately does not gate issue creation), still configurable
   * for a future creation-time gate and for symmetry with the data model. */
  from_state: string | null;
  to_state: string;
  is_active: boolean;
  approvers: TWorkflowTransitionApprover[];
  conditions: TWorkflowTransitionCondition[];
  actions: TWorkflowTransitionAction[];
  created_at: string;
  updated_at: string;
  created_by: string | null;
};

export type TWorkflowTransitionPayload = Partial<
  Pick<TWorkflowTransition, "issue_type" | "from_state" | "to_state" | "is_active" | "approvers" | "conditions" | "actions">
>;

export type TIssueTransitionApprovalDecision = "APPROVED" | "REJECTED";

export type TIssueTransitionApproval = {
  id: string;
  approval_request: string;
  approver: string | null;
  decision: TIssueTransitionApprovalDecision;
  comment: string;
  responded_at: string;
};

export type TIssueTransitionApprovalRequestStatus = "PENDING" | "APPROVED" | "REJECTED" | "CANCELLED";

export type TIssueTransitionApprovalRequest = {
  id: string;
  workspace_id: string;
  project_id: string;
  issue: string;
  transition: string;
  requested_by: string | null;
  status: TIssueTransitionApprovalRequestStatus;
  approvals: TIssueTransitionApproval[];
  created_at: string;
  updated_at: string;
};

export type TWorkflowTransitionAuditLogOutcome = "ALLOWED" | "DENIED" | "PENDING_APPROVAL";

export type TWorkflowTransitionAuditLog = {
  id: string;
  issue: string;
  transition: string | null;
  actor: string | null;
  from_state: string | null;
  to_state: string | null;
  outcome: TWorkflowTransitionAuditLogOutcome;
  denial_reason: string | null;
  created_at: string;
};

export type TWorkflowTransitionAuditLogParams = {
  cursor?: string;
  per_page?: number;
  issue_id?: string;
  outcome?: TWorkflowTransitionAuditLogOutcome;
  order_by?: string;
};

/** One entry per `State` in the issue's project - the shape
 * `IssueAllowedTransitionsEndpoint` returns. `allowed: false` covers BOTH a
 * hard deny (`reason_code` is one of `TRANSITION_NOT_IN_GRAPH`/
 * `ROLE_NOT_ALLOWED`/`CONDITION_NOT_MET`) and an approval-gated transition
 * (`reason_code === "APPROVAL_REQUIRED"`) - callers that need to tell these
 * apart (e.g. to keep an option selectable vs. fully disabled) must switch
 * on `reason_code`, not just `allowed`. */
export type TIssueAllowedTransitionReasonCode =
  | "APPROVAL_REQUIRED"
  | "TRANSITION_NOT_IN_GRAPH"
  | "ROLE_NOT_ALLOWED"
  | "CONDITION_NOT_MET";

export type TIssueAllowedTransitionEntry = {
  state_id: string;
  allowed: boolean;
  reason_code?: TIssueAllowedTransitionReasonCode;
  reason?: string;
};

/** The success-shaped-but-deferred response body a state-changing PATCH
 * returns (200, not the usual 204) when it hits a transition that needs
 * approval - see apps/api/plane/app/views/issue/base.py::partial_update and
 * apps/api/plane/api/views/issue.py::IssueDetailAPIEndpoint.patch. The
 * requested `state` was NOT written - only whatever other fields were
 * bundled into the same PATCH were. */
export type TIssueTransitionPendingApprovalResponse = {
  pending_approval: true;
  transition_id: string;
  approval_request_id: string;
};

/** The error body a state-changing PATCH returns (403) when the transition
 * is denied outright - never reaches the DB. */
export type TIssueTransitionDeniedError = {
  error_code: "TRANSITION_NOT_ALLOWED";
  reason: string;
};
